### PR TITLE
Зайцев Артем. Задача 2. Вариант 8. Обобщенная передача от одного всем (scatter).

### PR DESCRIPTION
**Pull request by aortemon**
### The problem 
Finding the minimum element of a vector
### Algorithm
1. Divide the vector into sub-vectors and distribute them among the threads.
2. Find the local minimums of each subvector.
3. Find the global minimum among local ones.
### Implementation
1. Signatures of the self-written `scatter(...)` and `MPI_Scatter(...)` are identical. 
2. If it's impossible to distribute the elements of the vector equally between the processes, the unallocated values from the end of the vector will be taken into account and processed in the root process. 
3. `class TestMPITaskParallel<T, func>(...)` is a template: when creating an instance, you must specify the type `T` of values in the vector and the function `func` that distributes data to processes. The tests consider the options `T ∈ {int, double}` and `func ∈ {scatter, MPI_Scatter}`.

### CODE REVIEW

**Review by Sean0450 — APPROVED** [review]

File: tasks/mpi/zaitsev_a_scatter/include/ops_mpi.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 Liolya Seledkina
+#pragma once
+
+#include <gtest/gtest.h>
```
Comment (Ksu04) [inline tasks/mpi/zaitsev_a_scatter/include/ops_mpi.hpp:4]: Delete this directive

Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/include/ops_mpi.hpp]: Removed

**Review by Ksu04 — APPROVED** [review]

File: tasks/mpi/zaitsev_a_scatter/perf_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  if (!testMpiTaskParallel->validation()) {
+    GTEST_SKIP();
+    return;
+  }
+
+  testMpiTaskParallel->pre_processing();
+  testMpiTaskParallel->run();
+  testMpiTaskParallel->post_processing();
+
+  // Create Perf attributes
```
[the remark this answers is not in the document]
Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/perf_tests/main.cpp]: Done

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Liolya Seledkina
```
[the remark this answers is not in the document]
Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp]: Done

**Review by PolinaSidorina — APPROVED** [review]

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
+#include <gtest/gtest.h>
```
Comment (allnes) [inline tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp:1]: where boost realization? please separate it into other directory

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+#include "mpi/zaitsev_a_scatter/include/ops_mpi.hpp"
+
+namespace zaitsev_a_scatter {
+std::vector<int> get_random_int_vector(int sz, int min, int max, int extrema = -100) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int> vec(sz);
+  for (int i = 0; i < sz; i++) vec[i] = min + gen() % (max - min + 1);
+  vec[0] = extrema;
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp:15]: Why do you need this line?

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  return vec;
+}
+
+std::vector<double> get_random_double_vector(int sz, double min, double max, double extrema) {
+  std::uniform_real_distribution<double> unif(min, max);
+  std::default_random_engine re;
+  std::vector<double> vec(sz);
+  for (int i = 0; i < sz; i++) vec[i] = unif(re);
+  vec[0] = extrema;
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp:24]: Same here

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  testMpiTaskParallel.pre_processing();
+  testMpiTaskParallel.run();
+  testMpiTaskParallel.post_processing();
+
+  if (world.rank() == 0) {
+    ASSERT_DOUBLE_EQ(extrema, reinterpret_cast<double*>(taskDataPar->outputs[0])[0]);
+  }
+}
+
+TEST(zaitsev_a_scatter, test__func_handwritten__dtype_int__size_1) { test_int<zaitsev_a_scatter::scatter>(1); }
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp:95]: Please, test empty arrays

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+  testMpiTaskParallel.pre_processing();
+  testMpiTaskParallel.run();
+  testMpiTaskParallel.post_processing();
+
+  if (world.rank() == 0) {
+    ASSERT_DOUBLE_EQ(extrema, reinterpret_cast<double*>(taskDataPar->outputs[0])[0]);
+  }
+}
+
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp:94]: Please add test on random numbers

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp (hunk 5 above)
Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp]: The 15th line quarantees the existence of the predefined lower extrema in a vector. This allows us to know the value of the desired minimum in advance and to skip the part of tests where an unknown minimum would be searched.

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp (hunk 4 above)
Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp]: Separated

File: tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp (hunk 5 above)
Comment (aortemon) [reply tasks/mpi/zaitsev_a_scatter/func_tests/main.cpp]: You are right, it was a mistake to use this approach. I discovered data duplication in streams which contributed to the incomplete coverage of the vector in the search for the minimum. I've fixed this issue. Tests you requested were added.

**Review by allnes — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]

**Pull request merged**